# Lecture 4: Pandas  & numpy continued 
- grab dataset from slack and save file in the same folder as this jupyter notebook 
-


In [3]:
#import libraries 
import pandas as pd
import numpy as np

C:\Users\chand\AppData\Local\Temp\ipykernel_26996\3900817540.py:2: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


In [5]:
#import data frame from file 
df = pd.read_csv('school_grades.csv')


In [6]:
#take a look at the dataframe 
print(df) #gives you the first five and last five in the dataframe; provides all columns 

     gender class group parental level of education         lunch  \
0      male     group A                 high school      standard   
1    female     group D            some high school  free/reduced   
2      male     group E                some college  free/reduced   
3      male     group B                 high school      standard   
4      male     group E          associate's degree      standard   
..      ...         ...                         ...           ...   
995    male     group C                 high school      standard   
996    male     group D          associate's degree  free/reduced   
997  female     group C            some high school  free/reduced   
998  female     group C                some college      standard   
999    male     group A                some college      standard   

    test preparation course  math score  reading score  writing score  
0                 completed          67             67             63  
1                      none

In [7]:
#get the top 10 rows 
print(df.head(10))

   gender class group parental level of education         lunch  \
0    male     group A                 high school      standard   
1  female     group D            some high school  free/reduced   
2    male     group E                some college  free/reduced   
3    male     group B                 high school      standard   
4    male     group E          associate's degree      standard   
5  female     group D                 high school      standard   
6  female     group A           bachelor's degree      standard   
7    male     group E                some college      standard   
8    male     group D                 high school      standard   
9    male     group C                some college  free/reduced   

  test preparation course  math score  reading score  writing score  
0               completed          67             67             63  
1                    none          40             59             55  
2                    none          59             60

In [8]:
#get the last 7 rows of the data set 
print(df.tail(7))

     gender class group parental level of education         lunch  \
993  female     group A          associate's degree      standard   
994    male     group E                 high school  free/reduced   
995    male     group C                 high school      standard   
996    male     group D          associate's degree  free/reduced   
997  female     group C            some high school  free/reduced   
998  female     group C                some college      standard   
999    male     group A                some college      standard   

    test preparation course  math score  reading score  writing score  
993                    none          58             54             58  
994               completed          86             82             75  
995                    none          73             70             65  
996               completed          85             91             92  
997                    none          32             35             41  
998            

In [9]:
#get a random sample! 
df.sample(5)

gender class group parental level of education         lunch  \
688  female     group B                some college  free/reduced   
509  female     group D                 high school      standard   
27     male     group E                 high school      standard   
651    male     group D          associate's degree  free/reduced   
748    male     group C           bachelor's degree      standard   

    test preparation course  math score  reading score  writing score  
688               completed          54             61             71  
509                    none          65             70             72  
27                     none          71             68             60  
651               completed          81             71             70  
748                    none          95             99             91

In [10]:
#get row and column count (rows, columns)
df.shape

(1000, 8)

In [11]:
print(f'our data frame has {df.shape[0]} rows and {df.shape[1]} columns')

our data frame has 1000 rows and 8 columns


In [12]:
#check for nulls 
df.isna().sum()

gender                         0
class group                    0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

In [13]:
#comprehensive summary of the dataset 
    #quick estimate of memory usage 
    #information about range 
    #info about the columns and type 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   class group                  1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


In [14]:
#to get deep and more exact memory usage 
df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   class group                  1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 303.4 KB


In [15]:
#an alternative to df.info is dtypes 
df.dtypes

gender                         object
class group                    object
parental level of education    object
lunch                          object
test preparation course        object
math score                      int64
reading score                   int64
writing score                   int64
dtype: object

**Observations** 
- we have 1000 rows and 8 columns 
    - 3 integer columns (int64) 
    - 5 string columns (object)
- there are no missing values in the dataset 
- index values from 0 to 999 
- real memory usage is 303.4 KB


In [16]:
#view the data using a specific range of rows 
df.loc[200:210]

gender class group parental level of education         lunch  \
200  female     group D          associate's degree      standard   
201    male     group E          associate's degree      standard   
202    male     group C                some college      standard   
203  female     group B            some high school      standard   
204  female     group D          associate's degree  free/reduced   
205    male     group B            some high school      standard   
206  female     group A             master's degree      standard   
207    male     group C                some college      standard   
208  female     group A                some college      standard   
209  female     group B          associate's degree      standard   
210  female     group D                 high school  free/reduced   

    test preparation course  math score  reading score  writing score  
200               completed          50             63             65  
201                    none          55             54             47  
202                    none          77             64             70  
203               completed          49             64             67  
204                    none          66             83             82  
205               completed          77             74             75  
206               completed          74             78             84  
207                    none          85             78             81  
208                    none          77             86             84  
209                    none          64             62             66  
210               completed          61             69             68

In [17]:
#view the data using a specific range of rows and columns 
df.iloc[200:210,5:8]

math score  reading score  writing score
200          50             63             65
201          55             54             47
202          77             64             70
203          49             64             67
204          66             83             82
205          77             74             75
206          74             78             84
207          85             78             81
208          77             86             84
209          64             62             66

In [18]:
#get specific column lists - with a sample of 10 
df[['gender','lunch','writing score']].sample(10)

gender         lunch  writing score
776    male  free/reduced             38
820  female      standard             74
48   female      standard             73
877    male  free/reduced             66
34   female      standard             89
570    male  free/reduced             60
821    male      standard             64
90     male  free/reduced             42
518    male  free/reduced             82
105  female      standard             82

In [19]:
df.iloc[100:111,[2,4,5]]

parental level of education test preparation course  math score
100                some college               completed          54
101                some college               completed          75
102                 high school                    none          79
103                 high school                    none          66
104                some college                    none          47
105           bachelor's degree                    none          70
106                 high school                    none          82
107           bachelor's degree                    none          56
108          associate's degree                    none          92
109                some college                    none          30
110                some college               completed          79

In [20]:
#drop columns and create a new data frame 
df2 = df.drop(columns=['gender','lunch'],axis=1)
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   class group                  1000 non-null   object
 1   parental level of education  1000 non-null   object
 2   test preparation course      1000 non-null   object
 3   math score                   1000 non-null   int64 
 4   reading score                1000 non-null   int64 
 5   writing score                1000 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 47.0+ KB


In [21]:
#changing data types
    #int64 = integer with 64 bits, which is a lot of memory! for a column with range 0 to 100 
    #change from int64 to int16 to same memory 

df['reading score'] = df['reading score'].astype('int16')

df.info(memory_usage='deep')

#membory went from 303.4 kb to 297.6kb , saving 5.8 kb of memory 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   class group                  1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int16 
 7   writing score                1000 non-null   int64 
dtypes: int16(1), int64(2), object(5)
memory usage: 297.6 KB


In [22]:
#get the sum of all numerical columns in a data frame 
df.sum(numeric_only=True)

math score       66396
reading score    69002
writing score    67738
dtype: int64

In [23]:
#sum of randomly selected rows of all numeric columns   
df.sample(10).sum(numeric_only=True)

math score       702
reading score    679
writing score    686
dtype: int64

In [24]:
df.max(numeric_only=True)

math score       100
reading score    100
writing score    100
dtype: int64

In [25]:
df.min(numeric_only=True)

math score       13
reading score    27
writing score    23
dtype: int64

In [26]:
#to get a full descriptive stats on numerical columns 
df.describe()

math score  reading score  writing score
count  1000.000000    1000.000000    1000.000000
mean     66.396000      69.002000      67.738000
std      15.402871      14.737272      15.600985
min      13.000000      27.000000      23.000000
25%      56.000000      60.000000      58.000000
50%      66.500000      70.000000      68.000000
75%      77.000000      79.000000      79.000000
max     100.000000     100.000000     100.000000

In [27]:
#mean of specific col 
df['math score'].mean()

66.396

In [28]:
#getting all the numerical columns 
num_col_list = df.select_dtypes(include='int').columns #note include and exclude can work here 
num_col_list

Index(['math score', 'writing score'], dtype='object')

In [30]:
#now i can take that list and convert them 
df[num_col_list] = df[num_col_list].astype('int16')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   class group                  1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int16 
 6   reading score                1000 non-null   int16 
 7   writing score                1000 non-null   int16 
dtypes: int16(3), object(5)
memory usage: 45.1+ KB


In [32]:
#renaming columns 
df2 = df.rename(columns={'reading score': 'reading_grade',
                         'math score': 'math_grade'})
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   class group                  1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math_grade                   1000 non-null   int16 
 6   reading_grade                1000 non-null   int16 
 7   writing score                1000 non-null   int16 
dtypes: int16(3), object(5)
memory usage: 45.1+ KB


In [34]:
df.columns = df.columns.str.replace(' ',"_")
df.columns

Index(['gender', 'class_group', 'parental_level_of_education', 'lunch',
       'test_preparation_course', 'math_score', 'reading_score',
       'writing_score'],
      dtype='object')

In [ ]:
#note use str.strip() to get rid of trailing empty spaces 

In [37]:
#mean across columns 
df['average_score'] = round(df.mean(numeric_only=True,axis=1),2)
df.head()

gender class_group parental_level_of_education         lunch  \
0    male     group A                 high school      standard   
1  female     group D            some high school  free/reduced   
2    male     group E                some college  free/reduced   
3    male     group B                 high school      standard   
4    male     group E          associate's degree      standard   

  test_preparation_course  math_score  reading_score  writing_score  \
0               completed          67             67             63   
1                    none          40             59             55   
2                    none          59             60             50   
3                    none          77             78             68   
4               completed          78             73             68   

   average_score  
0          65.67  
1          51.33  
2          56.33  
3          74.33  
4          73.00

### Aggregation

In [39]:
dfAveGender = df[['gender','math_score']].groupby('gender').mean()
dfAveGender

math_score
gender            
female   63.196687
male     69.384913

In [40]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 9 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   gender                       1000 non-null   object 
 1   class_group                  1000 non-null   object 
 2   parental_level_of_education  1000 non-null   object 
 3   lunch                        1000 non-null   object 
 4   test_preparation_course      1000 non-null   object 
 5   math_score                   1000 non-null   int16  
 6   reading_score                1000 non-null   int16  
 7   writing_score                1000 non-null   int16  
 8   average_score                1000 non-null   float64
dtypes: float64(1), int16(3), object(5)
memory usage: 52.9+ KB


In [48]:
#advanced aggregation 
#use average 
df.groupby('parental_level_of_education').agg({'math_score':'mean',
                                               'reading_score':'min'})

math_score  reading_score
parental_level_of_education                           
associate's degree            69.536946             34
bachelor's degree             71.491071             34
high school                   65.207921             28
master's degree               71.585714             40
some college                  65.297297             27
some high school              60.701571             31

In [50]:
#use a custom function 
    #e.g. diff between min and max 

def minmaxdiff(x): 
    return x.max() - x.min()


 

In [53]:
df.groupby('parental_level_of_education')[['math_score','reading_score']].agg(minmaxdiff)

math_score  reading_score
parental_level_of_education                           
associate's degree                   77             66
bachelor's degree                    65             66
high school                          71             72
master's degree                      62             60
some college                         87             73
some high school                     72             69